src/main/flight/alt_hold.c

```diff
@@ -0,0 +1,246 @@
+
+    if (newAltHoldEnabled && !altHoldState->altHoldEnabled)
+    {
+        altHoldReset(altHoldState);
+    }
+    if (!newAltHoldEnabled && altHoldState->altHoldEnabled) {
+        altHoldState->exitTime = millis();
+    }
+    altHoldState->altHoldEnabled = newAltHoldEnabled;
+
+    uint32_t currTime = millis();
+
+    if (newAltHoldEnabled) {
+        uint32_t timeSinceEnter = currTime - altHoldState->enterTime;
+        if (timeSinceEnter < ALTHOLD_ENTER_PERIOD) {
+            float delta = (float)timeSinceEnter / ALTHOLD_ENTER_PERIOD;
+            altHoldState->throttleFactor = MAX(delta, altHoldState->throttleFactor);
+        } else {
+            altHoldState->throttleFactor = 1.0f;
+        }
+        return;
+    }
+
+    if (altHoldState->exitTime == 0) {
+        altHoldState->throttleFactor = 0.0f;
+        return;
+    }
+
+    uint32_t timeSinceExit = currTime - altHoldState->exitTime;
+    if (timeSinceExit < ALTHOLD_MAX_EXIT_PERIOD) {
+        float delta = (float)timeSinceExit / ALTHOLD_MAX_EXIT_PERIOD;
+        altHoldState->throttleFactor = MIN(altHoldState->throttleFactor, 1.0f - delta);
+        return;
+    }
+
+    altHoldState->throttleFactor = 0.0f;
+}
+
+void altHoldUpdate(altHoldState_s* altHoldState)
+{
+    altHoldProcessTransitions(altHoldState);
+
+    float timeInterval = 1.0f / ALTHOLD_TASK_PERIOD;
+
+    float measuredAltitude = getCurrentAltitude(altHoldState);
+
+    t_fp_vector accelerationVector = {{
+        acc.accADC[X],
+        acc.accADC[Y],
+        acc.accADC[Z]
+    }};
+
+    imuTransformVectorBodyToEarth(&accelerationVector);
+
+    float measuredAccel = 9.8f * (accelerationVector.V.Z - acc.dev.acc_1G) / acc.dev.acc_1G;
+
+    DEBUG_SET(DEBUG_ALTHOLD, 0, (int16_t)(measuredAccel * 100.0f));
+
+    altHoldState->velocityEstimationAccel += measuredAccel * timeInterval;
+    altHoldState->velocityEstimationAccel *= 0.999f;
+
+    float currentVelocityEstimationAccel = altHoldState->velocityEstimationAccel - altHoldState->startVelocityEstimationAccel;
+    DEBUG_SET(DEBUG_ALTHOLD, 1, (int16_t)(100.0f * currentVelocityEstimationAccel));
+
+    altHoldState->measuredAltitude = measuredAltitude;
+    altHoldState->measuredAccel = measuredAccel;
+
+    if (altHoldState->altHoldEnabled) {
+        float velocityTarget = simplePidCalculate(&altHoldState->altPid, timeInterval, altHoldState->targetAltitude, altHoldState->measuredAltitude);
+
+        DEBUG_SET(DEBUG_ALTHOLD, 2, (int16_t)(100.0f * velocityTarget));
+
+        float velPidForce = simplePidCalculate(&altHoldState->velPid, timeInterval, velocityTarget, currentVelocityEstimationAccel);
+
+        DEBUG_SET(DEBUG_ALTHOLD, 3, (int16_t)(100.0f * velPidForce));
+
+        float newThrottle = velPidForce;
+
+        newThrottle = constrainf(newThrottle, 0.0f, 1.0f);
+        newThrottle = scaleRangef(newThrottle, 0.0f, 1.0f, 0.01f * altholdConfig()->minThrottle, 0.01f * altholdConfig()->maxThrottle);
+
+        altHoldState->throttle = newThrottle;
```

Comment:
just a general thought:
Here we are calculating throttle simply based on altitude changes.
When pilot steers left/right forward/back, it changes the roll/pitch angles, and that changes vertical thrust relative to the ground. Meaning the drone will dip down. Rather than waiting for PID to kick in, we can probably take into account the current roll/pitch. Something like this:
```
double effectiveTiltRadians = sqrt(sq(rollAngleRadians) + sq(pitchAngleRadians));
newThrottle = newThrottle / cos(effectiveTiltRadians);
```

Reply:
Leave this open for discussion. Like the idea.

Reply:
We do that in GPS rescue, when pitches forward.

Reply:
cos(effectiveTiltRadians) is directly available in orientation matrix (`getCosTiltAngle`)